tests/functional/eks/test_get_token.py

```diff
@@ -143,3 +144,105 @@ def test_url_different_partition(self):
             expected_endpoint='sts.cn-north-1.amazonaws.com.cn',
             expected_signing_region='cn-north-1'
         )
+
+    def test_api_version_discovery_deprecated(self):
+        os.environ["KUBERNETES_EXEC_INFO"] = '{"kind":"ExecCredential","apiVersion":"client.authentication.k8s.io/v1alpha1","spec":{"interactive":true}}'
+        cmd = 'eks get-token --cluster-name %s' % self.cluster_name
+        stdout, stderr, _ = self.run_cmd(cmd)
+        response = json.loads(stdout)
+
+        self.assertEqual(
+            response["apiVersion"],
+           "client.authentication.k8s.io/v1alpha1",
+        )
+
+        self.assertEqual(
+            stderr,
+           ("Kubeconfig user entry is using deprecated API "
+            "version client.authentication.k8s.io/v1alpha1. Run 'aws eks update-kubeconfig' to update\n")
+        )
+
+        del os.environ["KUBERNETES_EXEC_INFO"]
+
+    def test_api_version_discovery_malformed(self):
+        os.environ["KUBERNETES_EXEC_INFO"] = '{{"kind":"ExecCredential","apiVersion":"client.authentication.k8s.io/v1alpha1","spec":{"interactive":true}}'
+        cmd = 'eks get-token --cluster-name %s' % self.cluster_name
+        stdout, stderr, _ = self.run_cmd(cmd)
+        response = json.loads(stdout)
+
+        self.assertEqual(
+            response["apiVersion"],
+           "client.authentication.k8s.io/v1alpha1",
+        )
+
+        self.assertEqual(
+            stderr,
+           ("Error parsing KUBERNETES_EXEC_INFO, defaulting to client.authentication.k8s.io/v1alpha1. "
+            "This is likely a bug in your Kubernetes client. Please update your Kubernetes client.\n")
+        )
+
+        del os.environ["KUBERNETES_EXEC_INFO"]
+
+    def test_api_version_discovery_empty(self):
+        cmd = 'eks get-token --cluster-name %s' % self.cluster_name
+        stdout, stderr, _ = self.run_cmd(cmd)
+        response = json.loads(stdout)
+
+        self.assertEqual(
+            response["apiVersion"],
+           "client.authentication.k8s.io/v1alpha1",
+        )
+
+        self.assertEqual(
+            stderr,
+           ("Empty KUBERNETES_EXEC_INFO, defaulting to client.authentication.k8s.io/v1alpha1. "
+            "This is likely a bug in your Kubernetes client. Please update your Kubernetes client.\n")
+        )
+    def test_api_version_discovery_v1(self):
+        os.environ["KUBERNETES_EXEC_INFO"] = '{"kind":"ExecCredential","apiVersion":"client.authentication.k8s.io/v1","spec":{"interactive":true}}'
```

Comment:
It would be great if we had a helper function that sets the exec info based on the api version instead of having to duplicate the JSON document for each version. Specifically something like:
```python
def set_kubernetes_exec_info(self, api_version):
     # Set KUBERNETES_EXEC_INFO env var with default payload and provided api version
```
This will also make it easier to do bulk updates of the document used if needed in the future.

Reply:
We should consider just making it a context manager to do something like this. That also handles cleanup for us instead of using `del` everywhere.

```python
@contextlib.contextmanager
def set_environ(name, value):
    if value is not None:
        original_value = os.environ.get(name)
        os.environ[name] = value
    try:
        yield
    finally:
        if value is not None:
            if original_value is None:
                del os.environ[name]
            else:
                os.environ[name] = original_value
```

Reply:
The context manager seems at odds with the `self.environ` usage, I'll go for the helper function